scipy/special/_basic.py

```diff
@@ -3087,6 +3090,138 @@ def factorialk(n, k, exact=True):
     return _exact_factorialx_array(n, k=k)
 
 
+def stirling2(N, K, *, exact=True):
+    """Generate Stirling number(s) of the second kind.
+
+        Stirling numbers of the second kind count the number of ways to
+        partition a set with N elements into K non-empty subsets.
+
+        The values this function returns are calculated using a dynamic
+        program which avoids redundant computation across the subproblems
+        in the solution. For array-like input, this implementation also 
+        avoids redundant computation across the different Stirling number
+        calculations.
+
+        The numbers are sometimes denoted
+        .. math::
+
+            `{N \brace K}`;
+
+        see [GKP]_ for details. This is often expressed-verbally-as
+        "N subset K".
+
+        Parameters
+        ----------
+        N : int, ndarray
+            Number of things.
+        K : int, ndarray
+            Number of non-empty subsets taken.
+        exact : bool, optional
+            This keyword is reserved for a planned future implementation
+            that allows trading speed for accuracy.
+
+        Returns
+        -------
+        val : int, float, ndarray
+            The number of partitions.
+
+        See Also
+        --------
+        comb : The number of combinations of N things taken k at a time.
+
+        Notes
+        -----
+        - If N < 0, or K < 0, then 0 is returned.
+        - If K > N, then 0 is returned.
+        The output type will always be `int` or ndarray of `object`.
+        The input must contain either numpy or python integers otherwise a
+        TypeError is raised.
+
+        References
+        ----------
+         .. [GKP] R. L. Graham, D. E. Knuth and O. Patashnik, "Concrete
+        Mathematics: A Foundation for Computer Science," Addison-Wesley
+        Publishing Company, Boston, 1989. Chapter 6, page 258.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from scipy.special import stirling2
+        >>> k = np.array([3, -1, 3])
+        >>> n = np.array([10, 10, 9])
+        >>> stirling2(n, k)
+        array([9330, 0, 3025], dtype=object)
+
+    """
+    if exact:
+        # make a min-heap of unique (n,k) pairs
+        N, K = asarray(N), asarray(K)
+        if not (
+                np.issubdtype(N.dtype, np.integer)
+                or all(
+                    isinstance(n.take(0), int)
+                    for n in np.nditer(N, ['refs_ok'])
+                )
+        ):
+            raise TypeError("Argument `N` must contain only integers")
+        if not (
+                np.issubdtype(K.dtype, np.integer)
+                or all(
+                    isinstance(k.take(0), int)
+                    for k in np.nditer(K, ['refs_ok'])
+                )
+        ):
+            raise TypeError("Argument `K` contains non-integer type(s)")
```

Comment:
```suggestion
            raise TypeError("Argument `K` must contain only integers")
```
aligning raise messages